drv/sidecar-front-io-api/src/phy_smi.rs

```diff
@@ -0,0 +1,167 @@
+    pub fn phy_ctrl(&self) -> Result<u8, FpgaError> {
+        self.fpga.read(Addr::PHY_CTRL)
+    }
+
+    #[inline]
+    pub fn set_phy_ctrl(&self, val: u8) -> Result<(), FpgaError> {
+        self.fpga.write(WriteOp::Write, Addr::PHY_CTRL, val)
+    }
+
+    #[inline]
+    pub fn phy_power_enabled(&self) -> Result<bool, FpgaError> {
+        Ok((self.phy_ctrl()? & Reg::PHY_CTRL::EN) != 0)
+    }
+
+    #[inline]
+    pub fn set_phy_power_enabled(
+        &self,
+        enabled: bool,
+    ) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::from(enabled),
+            Addr::PHY_CTRL,
+            Reg::PHY_CTRL::EN,
+        )
+    }
+
+    #[inline]
+    pub fn set_phy_coma_mode(&self, asserted: bool) -> Result<(), FpgaError> {
+        self.fpga.write(
+            WriteOp::from(asserted),
+            Addr::PHY_CTRL,
+            Reg::PHY_CTRL::COMA_MODE,
+        )
+    }
+
+    #[inline]
+    pub fn phy_powered_up_and_ready(&self) -> Result<bool, FpgaError> {
+        let status: u8 = self.fpga.read(Addr::PHY_STATUS)?;
+        Ok((status & Reg::PHY_STATUS::READY) != 0)
+    }
+
+    #[inline]
+    pub fn smi_busy(&self) -> Result<bool, FpgaError> {
+        let status: u8 = self.fpga.read(Addr::PHY_SMI_STATUS)?;
+        Ok((status & Reg::PHY_SMI_STATUS::BUSY) != 0)
+    }
+
+    pub fn await_not_busy(&self) -> Result<(), FpgaError> {
+        while self.smi_busy()? {
+            if self.await_not_busy_sleep_for > 0 {
+                userlib::hl::sleep_for(self.await_not_busy_sleep_for);
+            }
+        }
+        Ok(())
+    }
+
+    fn read_raw<T: From<u16>>(
+        &self,
+        phy: u8,
+        reg: PhyRegisterAddress<T>,
+    ) -> Result<T, FpgaError> {
+        let request = SmiRequest {
+            rdata: U16::new(0),
+            wdata: U16::new(0),
+            phy,
+            reg: reg.addr,
+            ctrl: Reg::PHY_SMI_CTRL::START,
+        };
+        self.await_not_busy()?;
+        self.fpga
+            .write(WriteOp::Write, Addr::PHY_SMI_RDATA_H, request)?;
+
+        self.await_not_busy()?;
+        let v = u16::from_be(self.fpga.read(Addr::PHY_SMI_RDATA_H)?);
+
+        Ok(v.into())
+    }
+
+    fn write_raw<T>(
+        &self,
+        phy: u8,
+        reg: PhyRegisterAddress<T>,
+        value: T,
+    ) -> Result<(), FpgaError>
+    where
+        u16: From<T>,
+        T: From<u16> + Clone,
+    {
+        let request = SmiRequest {
+            rdata: U16::new(0),
+            wdata: U16::new(value.into()),
+            phy,
+            reg: reg.addr,
+            ctrl: Reg::PHY_SMI_CTRL::RW | Reg::PHY_SMI_CTRL::START,
+        };
+        self.await_not_busy()?;
+        self.fpga
+            .write(WriteOp::Write, Addr::PHY_SMI_RDATA_H, request)
+    }
+}
+
+impl PhyRw for PhySmi {
+    #[inline]
+    fn read_raw<T: From<u16>>(
+        &self,
+        phy: u8,
+        reg: PhyRegisterAddress<T>,
+    ) -> Result<T, VscError> {
+        self.read_raw(phy, reg)
```

Comment:
It's not obvious that this calls `PhySmi::read_raw`, instead of recursion forever – maybe rename the inner functions to `read/write_raw_inner`?